components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```diff
@@ -0,0 +1,630 @@
+            if (StringUtils.isBlank(logoutToken)) {
+                throw new LogoutClientException(ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getCode(),
+                        ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getMessage());
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Handling the OIDC federated IdP Initiated logout request for the obtained logout token: " +
+                        logoutToken);
+            }
+            // Get the claim set from the logout token.
+            SignedJWT signedJWT = SignedJWT.parse(logoutToken);
+            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();
+            // Check for the iss value in claim set.
+            isIssuerClaimExists(claimsSet);
+            // Get the identity provider for the issuer of the logout token.
+            String tenantDomain = logoutRequest.getTenantDomain();
+            IdentityProvider identityProvider = getIdentityProvider(claimsSet.getIssuer(), tenantDomain);
+
+            validateLogoutToken(signedJWT, identityProvider);
+            // Retrieve the federated user id from the IDN_AUTH_USER table.
+            String sub = claimsSet.getSubject();
+            String userId = getUserId(tenantDomain, sub, identityProvider);
+            if (log.isDebugEnabled()) {
+                log.debug("Trying OIDC federated identity provider initiated logout for the user: " + sub);
+            }
+
+            // Check for sid or sub claims in the logout token.
+            if (isSidClaimExists(claimsSet)) {
+                // Check whether the sid has an entry in the federated authentication session details table.
+                return logoutUsingSid((String) claimsSet.getClaim(OIDCAuthenticatorConstants.Claim.SID), userId);
+            }
+            // Check whether the sub claim is available in the logout token.
+            if (log.isDebugEnabled()) {
+                log.debug("No sid presented in the logout token of the federated idp initiated logout request. Using " +
+                        "sub claim to terminate the sessions for tenant domain: " + tenantDomain);
+            }
+            // Check whether the sub claim has a valid user in the IS.
+            return logoutUsingSub(userId);
+
+        } catch (ParseException e) {
+            throw handleLogoutClientException(ErrorMessages.LOGOUT_TOKEN_PARSING_FAILURE, e);
+        }
+    }
+
+    /**
+     * Terminate the session related to the sid value of the logout token.
+     *
+     * @param sid - sid claim included in the logout token.
+     * @return
+     * @throws LogoutServerException
+     */
+    private LogoutResponse.LogoutResponseBuilder logoutUsingSid(String sid, String userId)
+            throws LogoutServerException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Trying federated IdP initiated logout using sid: " + sid);
+        }
+        String sessionId = getSessionIdFromSid(sid);
+        if (sessionId == null) {
```

Comment:
Do we need any instance where sessionId is null. It is already handled in `getSessionIdFromSid()` method.